### ISSUE TITLE
Fuse reshape into pointwise kernels (might be MLIR-only?)

### ISSUE THREAD

**Issue by krzysz00** [ROCmSoftwarePlatform/AMDMIGraphX#1844]
I was looking through the compile trace on `distilgpt2_1_fp16_gpu.onnx` and I noticed that there were cases where there was a pointwise op, a `reshape`, and then a long string of pointwise operations, but that the reshape wasn't fused into the pointwise module.

Since, as far as I know, `reshape` is effectively a pointwise operation (it just affects your indexing in subsequent operations), it should be fused in to pointwise modules.

At the very least, I'm pretty sure MLIR fusions can handle this case.

**Comment by krzysz00** [comment]
In other words, (using some pseudocode)
```
%0 = dot(...) : 64x64
%1 = add(%0, %arg1) : 64x64
%2 = reshape(%1) : 1x64x64
%3 = mul(%2, %arg2) : 1x64x64
...
return
```
should be one pointwise module, but it's currently two (if I read the results of the compile correctly)
From what I can tell, this is costing us a substantial fusion opportunity (at least when MLIR is enabled)

**Comment by krzysz00** [comment]
Yeah, I think doing this in fuse_mlir by matching on `@x = gpu::mlir_op(...); @y = reshape(@x); @z = pointwise(@y)` might be the right call here.

Who'd be a good fit for writing that pass improvement?

**Comment by krzysz00** [comment]
I've found a cleaner example of why we need this 

```
@311 = gpu::code_object[code_object=6312,symbol_name=mlir_dot,global=3072,local=128,](@309,@308,@302,@310) -> half_type, {1, 3072}, {3072, 1}, target_id=0
@312 = reshape[dims={1, 1, 3072}](@311) -> half_type, {1, 1, 3072}, {3072, 3072, 1}, target_id=0
@313 = load[offset=1536,end=7680](@1) -> half_type, {1, 1, 3072}, {3072, 3072, 1}, target_id=0
@314 = gpu::code_object[code_object=13680,symbol_name=pow_mul_add_mul_tanh_add_mul_mul_kernel,global=1536,local=1024,](@312,@313) -> half_type, {1, 1, 3072}, {3072, 3072, 1}, target_id=0
```

from 
`env MIGRAPHX_ENABLE_MLIR=1 MIGRAPHX_TRACE_COMPILE=0 MIGRAPHX_TRACE_MLIR=1 ./bin/driver run --onnx /data/onnx_models/distilgpt2_1_fp16_gpu.onnx --fill1 input_ids 2>&1 | less`

@pfultz2

**Comment by krzysz00** [comment]
There's also stuff like
```
@1074 = gpu::code_object[code_object=6056,symbol_name=mlir_dot,global=5120,local=64,](@1073,@1071,@1062,@1072) -> float_type, {1, 5120}, {5120, 1}, target_id=0
@1075 = load[offset=5120,end=25600](@1) -> float_type, {1, 1, 5120}, {5120, 5120, 1}, target_id=0
@1076 = reshape[dims={1, 1, 5120}](@1074) -> float_type, {1, 1, 5120}, {5120, 5120, 1}, target_id=0
@1077 = gpu::code_object[code_object=13680,symbol_name=pow_mul_add_mul_tanh_add_mul_mul_kernel,global=2560,local=1024,](@1076,@1075) -> float_type, {1, 1, 5120}, {5120, 5120, 1}, target_id=0
@1078 = multibroadcast[out_lens={1, 1280},out_dyn_dims={}](@1059) -> float_type, {1, 1280}, {0, 1}, target_id=0
@1079 = load[offset=25600,end=30720](@1) -> float_type, {1, 1280}, {1280, 1}, target_id=0
@1080 = gpu::code_object[code_object=6056,symbol_name=mlir_dot,global=1280,local=64,](@1078,@1077,@1060,@1079) -> float_type, {1, 1280}, {1280, 1}, target_id=0
@1081 = hip::hip_copy_literal[id=main:@literal:183] -> float_type, {1280}, {1}, target_id=0
@1082 = hip::hip_copy_literal[id=main:@literal:182] -> float_type, {1280}, {1}, target_id=0
@1083 = reshape[dims={1, 1, 1280}](@1080) -> float_type, {1, 1, 1280}, {1280, 1280, 1}, target_id=0
@1083 = reshape[dims={1, 1, 1280}](@1080) -> float_type, {1, 1, 1280}, {1280, 1280, 1}, target_id=0
@1084 = load[offset=30720,end=35840](@1) -> float_type, {1, 1, 1280}, {1280, 1280, 1}, target_id=0
@1085 = gpu::code_object[code_object=13624,symbol_name=add_kernel,global=640,local=1024,](@1067,@1083,@1084) -> float_type, {1, 1, 1280}, {1280, 1280, 1}, target_id=0
@1086 = multibroadcast[out_lens={1, 1, 1280},out_dyn_dims={}](@1082) -> float_type, {1, 1, 1280}, {0, 0, 1}, target_id=0
@1087 = multibroadcast[out_lens={1, 1, 1280},out_dyn_dims={}](@1081) -> float_type, {1, 1, 1280}, {0, 0, 1}, target_id=0
@1088 = load[offset=15360,end=20480](@1) -> float_type, {1, 1, 1280}, {1280, 1280, 1}, target_id=0
@1089 = gpu::code_object[code_object=13648,symbol_name=layernorm_mul_add_kernel,global=256,local=256,](@1085,@1087,@1086,@1088) -> float_type, {1, 1, 1280}, {1280, 1280, 1}, target_id=0
```

which looks like more fusion opportunities blocked by the lack of a reshape fusion.

There's also potentially another ticket here about fusing _slices_ into MLIR to lower the number of copies, but that's maybe something else.
